src/query/atomWithInfiniteQuery.ts

```diff
@@ -50,14 +50,23 @@ export function atomWithInfiniteQuery<
       let settlePromise:
         | ((data: InfiniteData<TData> | null, err?: TError) => void)
         | null = null
-      const getInitialData = () =>
-        typeof options.initialData === 'function'
-          ? (
-              options.initialData as InitialDataFunction<
-                InfiniteData<TQueryData>
-              >
-            )()
-          : options.initialData
+
+      const getInitialData = () => {
+        let data: InfiniteData<TQueryData> | InfiniteData<TData> | undefined =
+          queryClient.getQueryData<InfiniteData<TData>>(options.queryKey)
+
+        if (!data && options.initialData)
```

Comment:
Doesn't this have to be `data === undefined && ...`? For data can be null or 0?

Reply:
great catch

Reply:
fixed in https://github.com/pmndrs/jotai/pull/653/commits/d2599e20e2eea00b1b23f100b725da5b463236ce